### PR TITLE
z_listreceivedbyaddress: reject UA component addr (#5537)

### LINKED ISSUE #5537: z_listreceivedbyaddress should return an error if it is passed a bare UA component address.
This was left over from #5467:
- [ ] When given a P2PKH, P2SH, or Sapling address, this API should return an error if that address corresponds to a component of a unified address managed by this wallet.

### PR DESCRIPTION
**Pull request by LarryRuane**
Closes #5537

### CODE REVIEW

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -3470,6 +3470,29 @@ UniValue z_listreceivedbyaddress(const UniValue& params, bool fHelp)
...
+        },
+        [&](const libzcash::SproutPaymentAddress& addr) {
+            // A unified address can't contain a Sprout receiver.
+            return false;
+        },
+        [&](const libzcash::UnifiedAddress& addr) {
+            return false;
+        }
+    }, decoded.value())) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "z_listreceivedbyaddress does not allow a unified address argument");
```
Comment (nuttycom) [inline src/wallet/rpcwallet.cpp:3493]:
```suggestion
        throw JSONRPCError(RPC_INVALID_PARAMETER, "The provided address is a bare receiver from a Unified Address in this wallet. Provide the full UA instead.");
```

File: src/wallet/rpcwallet.cpp [hunk 2]
```diff
@@ -3470,6 +3470,29 @@ UniValue z_listreceivedbyaddress(const UniValue& params, bool fHelp)
...
+        },
+        [&](const libzcash::SaplingPaymentAddress& addr) {
+            return pwalletMain->FindUnifiedAddressByReceiver(addr).has_value();
+        },
+        [&](const libzcash::SproutPaymentAddress& addr) {
+            // A unified address can't contain a Sprout receiver.
+            return false;
+        },
+        [&](const libzcash::UnifiedAddress& addr) {
+            return false;
```
Comment (str4d) [inline src/wallet/rpcwallet.cpp:3490]:
```suggestion
            // We allow unified addresses themselves.
            return false;
```

File: qa/rpc-tests/wallet_listreceived.py [hunk 3]
```diff
@@ -384,6 +384,15 @@ def test_received_sapling(self, height):
         assert_equal(len(receivers), 2)
         assert 'transparent' in receivers
         assert 'sapling' in receivers
+        assert_raises_message(
+            JSONRPCException,
+            "z_listreceivedbyaddress does not allow a unified address argument",
+            node.z_listreceivedbyaddress, receivers['transparent'], 0)
+        assert_raises_message(
+            JSONRPCException,
+            "z_listreceivedbyaddress does not allow a unified address argument",
```
Comment (nuttycom) [inline qa/rpc-tests/wallet_listreceived.py:393]:
```suggestion
            "The provided address is a bare receiver from a Unified Address in this wallet. Provide the full UA instead.",
            node.z_listreceivedbyaddress, receivers['transparent'], 0)
        assert_raises_message(
            JSONRPCException,
            "The provided address is a bare receiver from a Unified Address in this wallet. Provide the full UA instead.",
```

**Review by nuttycom — CHANGES_REQUESTED** [review]
Looks good, one minor change is needed to how the error is reported.

**Review by nuttycom — APPROVED** [review]
utACK

**Review by str4d — APPROVED** [review]

File: src/wallet/rpcwallet.cpp [hunk 4]
```diff
@@ -3470,6 +3470,30 @@ UniValue z_listreceivedbyaddress(const UniValue& params, bool fHelp)
         throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid zaddr.");
     }
 
+    // A non-unified address argument that is a receiver within a
+    // unified address known to this wallet is not allowed.
```
Comment (daira) [inline src/wallet/rpcwallet.cpp:3474]: Wait, why not? This seems like an odd exception. What if the provider of the bare address doesn't know the full UA? Why should an address that happens to be part of a UA known to the wallet behave differently to any other bare address? I would definitely have pushed back on this if I hadn't been distracted by unrelated stuff yesterday.

**Pull request merged**